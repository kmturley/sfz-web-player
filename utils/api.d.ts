declare function get(url: string): Promise<string>;
declare function getJSON(url: string): Promise<any>;
declare function getRaw(url: string): Promise<ArrayBuffer>;
declare function getXML(url: string): Promise<object>;
export { get, getJSON, getRaw, getXML };
