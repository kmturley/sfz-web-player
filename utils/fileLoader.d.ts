import { FileGitHubItem, FileItem, FilesMap, FilesNested } from "../types/files";
declare class FileLoader {
    private branch;
    private files;
    private filesNested;
    private directory;
    getFiles(): FilesMap;
    loadFileLocal(file: File): Promise<FileItem>;
    loadFileRemote(file: FileGitHubItem): Promise<FileItem>;
    loadDirectoryLocal(blobs: File[]): Promise<{
        branch: string;
        directory: string;
        files: FilesMap;
        filesNested: FilesNested;
    }>;
    loadDirectoryRemote(data: any): Promise<{
        branch: string;
        directory: string;
        files: FilesMap;
        filesNested: FilesNested;
    }>;
}
export default FileLoader;
