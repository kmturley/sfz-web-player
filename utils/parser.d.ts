import { AudioSfz, AudioSfzVariables } from '../types/audio';
import FileLoader from './fileLoader';
declare function parseSfz(prefix: string, contents: string): Promise<any>;
declare function processDirective(input: string): RegExpMatchArray | [];
declare function processHeader(input: string): RegExpMatchArray | [];
declare function processOpcode(input: string): any;
declare function processVariables(input: string, vars: AudioSfzVariables): string;
declare function flattenSfzObject(sfzObject: AudioSfz): any;
declare function setParserLoader(fileLoader: FileLoader): void;
export { flattenSfzObject, parseSfz, processDirective, processHeader, processOpcode, processVariables, setParserLoader, };
