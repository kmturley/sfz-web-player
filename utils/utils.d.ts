declare function encodeHashes(item: string): string;
declare function midiNameToNum(name: string | number): any;
declare function pathDir(item: string, separator?: string): string;
declare function pathExt(item: string): string;
declare function pathJoin(...segments: any): string;
declare function pathRoot(item: string, separator?: string): string;
declare function pathSubDir(item: string, dir: string): string;
export { encodeHashes, midiNameToNum, pathDir, pathExt, pathJoin, pathRoot, pathSubDir };
