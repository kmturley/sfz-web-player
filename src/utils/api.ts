import { xml2js } from "xml-js";

async function get(url: string): Promise<string> {
  console.log("⤓", url);
  return fetch(url).then((res: any) => res.text());
}

async function getJSON(url: string): Promise<any> {
  console.log("⤓", url);
  return fetch(url).then((res: any) => res.json());
}

async function getRaw(url: string): Promise<ArrayBuffer> {
  console.log("⤓", url);
  return fetch(url).then((res: any) => res.arrayBuffer());
}

async function getXML(url: string): Promise<object> {
  console.log("⤓", url);
  return fetch(url).then(async (res: any) => xml2js(await res.text()));
}

export { get, getJSON, getRaw, getXML };
