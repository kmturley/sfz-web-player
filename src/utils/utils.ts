function encodeHashes(item: string) {
  return item.replace(/#/g, encodeURIComponent('#'));
}

function pathDir(item: string, separator: string = '/'): string {
  return item.substring(0, item.lastIndexOf(separator) + 1);
}

function pathExt(item: string): string {
  return item.substring(item.lastIndexOf('.') + 1);
}

function pathJoin(...segments: any) {
  const parts = segments.reduce((partItems: any, segment: any) => {
    // Remove leading slashes from non-first part.
    if (partItems.length > 0) {
      segment = segment.replace(/^\//, '');
    }
    // Remove trailing slashes.
    segment = segment.replace(/\/$/, '');
    return partItems.concat(segment.split('/'));
  }, []);
  const resultParts = [];
  for (const part of parts) {
    if (part === '.') {
      continue;
    }
    if (part === '..') {
      const partRemoved: string = resultParts.pop();
      if (partRemoved === '') resultParts.pop();
      continue;
    }
    resultParts.push(part);
  }
  return resultParts.join('/');
}

function pathRoot(item: string, separator: string = '/'): string {
  return item.substring(0, item.indexOf(separator) + 1);
}

function pathSubDir(item: string, dir: string): string {
  return item.replace(dir, '');
}

export { encodeHashes, pathDir, pathExt, pathJoin, pathRoot, pathSubDir };
