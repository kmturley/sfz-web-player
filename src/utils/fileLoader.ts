import { FileWithDirectoryAndFileHandle } from "browser-fs-access";
import { FileLocal, FileRemote, FilesMap, FilesTree } from "../types/files";
import { get } from "./api";
import { pathDir, pathExt, pathRoot, pathSubDir } from "./utils";

class FileLoader {
  files: FilesMap = {};
  filesTree: FilesTree = {};
  root: string = "";

  addDirectory(files: string[] | FileWithDirectoryAndFileHandle[]) {
    this.resetFiles();
    files.forEach((file: string | FileWithDirectoryAndFileHandle) =>
      this.addFile(file)
    );
  }

  addFile(file: string | FileWithDirectoryAndFileHandle) {
    let item: FileLocal | FileRemote;
    if (typeof file === "string") {
      if (file === this.root) return;
      item = {
        ext: pathExt(file),
        contents: null,
        path: decodeURI(file),
      };
    } else {
      item = {
        ext: pathExt(file.webkitRelativePath),
        contents: null,
        path: decodeURI(file.webkitRelativePath),
        handle: file,
      };
    }
    const fileKey: string = pathSubDir(item.path, this.root);
    this.files[fileKey] = item;
    fileKey
      .split("/")
      .reduce((o: any, k: string) => (o[k] = o[k] || {}), this.filesTree);
    return item;
  }

  async loadFile(file: string | FileWithDirectoryAndFileHandle) {
    if (typeof file === "string") {
      return {
        ext: pathExt(file),
        contents: await get(file),
        path: decodeURI(file),
      } as FileRemote;
    } else {
      return {
        ext: pathExt(file.webkitRelativePath),
        contents: await file.text(),
        path: file.webkitRelativePath,
      } as FileLocal;
    }
  }

  async getFile(file: string | FileLocal | FileRemote | undefined) {
    if (!file) return;
    if (typeof file === "string") {
      const fileKey: string = pathSubDir(file, this.root);
      if (this.files[fileKey]) {
        file = this.files[fileKey];
      } else {
        file = await this.loadFile(file);
      }
    }
    if (!file.contents) {
      const fileKey: string = pathSubDir(file.path, this.root);
      if ("handle" in file) {
        file = await this.loadFile(file.handle);
        this.files[fileKey] = file;
      } else {
        file = await this.loadFile(file.path);
        this.files[fileKey] = file;
      }
    }
    return file;
  }

  setRoot(dir: string) {
    console.log("setRoot", dir);
    this.root = dir;
  }

  resetFiles() {
    this.files = {};
    this.filesTree = {};
  }
}

export default FileLoader;
