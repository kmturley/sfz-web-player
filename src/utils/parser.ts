import { AudioSfz, AudioSfzOpcode, AudioSfzOpcodeObj, AudioSfzOpcodes, AudioSfzVariables } from '../types/audio';
import { FileLocal, FileRemote } from '../types/files';
import FileLoader from './fileLoader';
import { midiNameToNum, pathJoin } from './utils';

let loader: FileLoader;

const DEBUG: boolean = false;
const skipCharacters: string[] = [' ', '\t', '\r', '\n'];
const endCharacters: string[] = ['>', '\r', '\n'];
const variables: any = {};

async function parseSfz(prefix: string, contents: string) {
  let elements: any[] = [];
  let element: any = {};
  for (let i: number = 0; i < contents.length; i++) {
    const char: string = contents.charAt(i);
    if (skipCharacters.includes(char)) continue; // skip character
    const iEnd: number = findEnd(contents, i);
    let line: string = contents.slice(i, iEnd);
    if (char === '/') {
      // do nothing
    } else if (char === '#') {
      const matches: string[] = processDirective(line);
      if (matches[0] === 'include') {
        const includeVal: any = await loadParseSfz(prefix, matches[1]);
        if (includeVal.elements) {
          element.elements = element.elements.concat(includeVal.elements);
        } else {
          elements = elements.concat(includeVal);
        }
        if (DEBUG) console.log('include', matches[1], JSON.stringify(includeVal));
      } else if (matches[0] === 'define') {
        variables[matches[1]] = matches[2];
        if (DEBUG) console.log('define', matches[1], variables[matches[1]]);
      }
    } else if (char === '<') {
      const matches: string[] = processHeader(line);
      element = {
        type: 'element',
        name: matches[0],
        elements: [],
      };
      elements.push(element);
      if (DEBUG) console.log(`<${element.name}>`);
    } else {
      if (line.includes('$')) line = processVariables(line, variables);
      if (!element.elements) {
        element.elements = [];
      }
      const opcodes: AudioSfzOpcode[] = processOpcode(line);
      opcodes.forEach((opcode: AudioSfzOpcode) => {
        element.elements.push({
          type: 'element',
          name: 'opcode',
          attributes: opcode,
        });
      });
      if (DEBUG) console.log(line, opcodes);
    }
    i = iEnd;
  }
  if (!element.type) return element;
  return elements;
}

function containsHeader(data: any) {
  if (data.region) return 'region';
  if (data.group) return 'group';
  if (data.control) return 'control';
  if (data.global) return 'global';
  if (data.curve) return 'curve';
  if (data.effect) return 'effect';
  if (data.master) return 'master';
  if (data.midi) return 'midi';
  if (data.sample) return 'sample';
  return undefined;
}

async function loadParseSfz(prefix: string, suffix: string) {
  const pathJoined: string = pathJoin(prefix, suffix);
  const fileRef: FileLocal | FileRemote | undefined = loader.files[pathJoined];
  const file: FileLocal | FileRemote | undefined = await loader.getFile(fileRef || pathJoined);
  return await parseSfz(prefix, file?.contents);
}

function processDirective(input: string) {
  return input.match(/[^# "]+/g) || [];
}

function processHeader(input: string) {
  return input.match(/[^< >]+/g) || [];
}

function processOpcode(input: string) {
  const output: AudioSfzOpcode[] = [];
  const labels: string[] = input.match(/\w+(?==)/g) || [];
  const values: string[] = input.split(/\w+(?==)/g) || [];
  values.forEach((val: string) => {
    if (!val.length) return;
    output.push({
      name: labels[output.length],
      value: val.trim().replace(/[='"]/g, ''),
    });
  });
  return output;
}

function processOpcodeObject(input: string) {
  const output: AudioSfzOpcodeObj = {};
  const labels: string[] = input.match(/\w+(?==)/g) || [];
  const values: string[] = input.split(/\w+(?==)/g) || [];
  values.forEach((val: string) => {
    if (!val.length) return;
    const opcodeName: string = labels[Object.keys(output).length];
    const opcodeValue: string = val.trim().replace(/[='"]/g, '');
    if (!isNaN(opcodeValue as any)) {
      output[opcodeName] = Number(opcodeValue);
    } else {
      output[opcodeName] = opcodeValue;
    }
  });
  return output;
}

function processVariables(input: string, vars: AudioSfzVariables) {
  const list: string = Object.keys(vars)
    .map((key) => '\\' + key)
    .join('|');
  const regEx: RegExp = new RegExp(list, 'g');
  return input.replace(regEx, (matched: string) => {
    return vars[matched];
  });
}

function flattenSfzObject(sfzObject: AudioSfzOpcodes, groupIndex = 0) {
  const keys: any = {};
  const groupObj: any = opcodesToObject(sfzObject.opcode);
  sfzObject.region?.forEach((region: AudioSfzOpcodes) => {
    const regionObj: any = opcodesToObject(region.opcode);
    const mergedObj: any = { ...groupObj, ...regionObj };
    const start: number = midiNameToNum(regionObj.lokey || regionObj.key);
    const end: number = midiNameToNum(regionObj.hikey || regionObj.key);
    if (start === 0 && end === 0) return;
    for (let i = start; i <= end; i++) {
      if (!keys[i]) keys[i] = [];
      keys[i].push(mergedObj);
    }
  });
  return keys;
}

function opcodesToObject(opcodes: AudioSfzOpcode[]) {
  const properties: any = {};
  opcodes.forEach((opcode: AudioSfzOpcode) => {
    if (!isNaN(opcode.value as any)) {
      properties[opcode.name] = Number(opcode.value);
    } else {
      properties[opcode.name] = opcode.value;
    }
  });
  return properties;
}

function findEnd(contents: string, startAt: number) {
  const isComment: boolean = contents.charAt(startAt) === '/' && contents.charAt(startAt + 1) === '/';
  for (let index: number = startAt; index < contents.length; index++) {
    const char: string = contents.charAt(index);
    if (isComment && char === '>') continue;
    if (endCharacters.includes(char)) return index;
    if (index > startAt + 1 && char === '/' && contents.charAt(index + 1) === '/') return index;
  }
  return contents.length;
}

function setParserLoader(fileLoader: FileLoader) {
  loader = fileLoader;
}

export {
  findEnd,
  flattenSfzObject,
  parseSfz,
  processDirective,
  processHeader,
  processOpcode,
  processOpcodeObject,
  processVariables,
  setParserLoader,
};
