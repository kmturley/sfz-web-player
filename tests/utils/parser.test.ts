import { get } from '../../src/utils/api';
import { parseSfz, processDirective, processHeader, processOpcode, processVariables } from '../../src/utils/parser';
import { encodeHashes, pathDir } from '../../src/utils/utils';
import 'whatwg-fetch';

const sfzFiles: string[] = [
  'https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/01%20-%20amp%20lfo%20freq.sfz',
  'https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/02%20-%20amp%20lfo%20freq%20cc1%20and%20after.sfz',
  'https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/03%20-%20amp%20lfo%20depth.sfz',
  'https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/04%20-%20amp%20lfo%20depth%20on%20cc.sfz',
  'https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/05%20-%20amp%20lfo%20depth%20on%20channel%20aftertouch.sfz',
  'https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/06%20-%20amp%20lfo%20delay.sfz',
  'https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/07%20-%20amp%20lfo%20fade.sfz',
];

test.each(sfzFiles)('parseSfz %p', async (sfzFile: string) => {
  const prefix: string = pathDir(sfzFile);
  const contents: string = await get(encodeHashes(sfzFile));
  expect(await parseSfz(prefix, contents, true)).toMatchSnapshot();
});

test('processDirective', () => {
  expect(processDirective('#include "green/stac_tp.sfz"')).toEqual(['include', 'green/stac_tp.sfz']);
  expect(processDirective('#define $KICKKEY 36')).toEqual(['define', '$KICKKEY', '36']);
});

test('processHeader', () => {
  expect(processHeader('<region')).toEqual(['region']);
  expect(processHeader('< region')).toEqual(['region']);
  expect(processHeader('<region>')).toEqual(['region']);
  expect(processHeader('<region >')).toEqual(['region']);
  expect(processHeader('region>')).toEqual(['region']);
});

test('processOpcode', () => {
  expect(processOpcode('seq_position=3')).toEqual({ seq_position: 3 });
  expect(processOpcode('seq_position=3 pitch_keycenter=50')).toEqual({ seq_position: 3, pitch_keycenter: 50 });
  expect(processOpcode('region_label=01 sample=harmLA0.$EXT')).toEqual({
    region_label: 1,
    sample: 'harmLA0.$EXT',
  });
  expect(processOpcode('label_cc27="Release vol"')).toEqual({ label_cc27: 'Release vol' });
  expect(processOpcode('label_cc27=Release vol')).toEqual({ label_cc27: 'Release vol' });
  expect(processOpcode('apple=An Apple banana=\'A Banana\' carrot="A Carrot"')).toEqual({
    apple: 'An Apple',
    banana: 'A Banana',
    carrot: 'A Carrot',
  });
});

test('processVariables', () => {
  expect(processVariables('sample=harmLA0.$EXT', { $EXT: 'flac' })).toEqual('sample=harmLA0.flac');
});
