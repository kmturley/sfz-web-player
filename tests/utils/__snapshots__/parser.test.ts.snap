// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseSfz "https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/01%20-%20amp%20lfo%20freq.sfz" 1`] = `
{
  "global": [
    {
      "group": [
        {
          "region": [
            {
              "amplfo_depth": 6,
              "amplfo_freq": 1,
              "loopend": 56728,
              "loopmode": "loop_continuous",
              "loopstart": 12629,
              "sample": "..\\..\\samples\\440.wav",
              "volume": -10,
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parseSfz "https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/02%20-%20amp%20lfo%20freq%20cc1%20and%20after.sfz" 1`] = `
{
  "global": [
    {
      "group": [
        {
          "region": [
            {
              "amplfo_depth": 6,
              "amplfo_freq": 2,
              "amplfo_freqcc1": -2,
              "amplfo_freqchanaft": 2,
              "loopend": 56728,
              "loopmode": "loop_continuous",
              "loopstart": 12629,
              "sample": "..\\..\\samples\\440.wav",
              "volume": -10,
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parseSfz "https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/03%20-%20amp%20lfo%20depth.sfz" 1`] = `
{
  "global": [
    {
      "group": [
        {
          "region": [
            {
              "amplfo_depth": 12,
              "amplfo_freq": 3,
              "loopend": 56728,
              "loopmode": "loop_continuous",
              "loopstart": 12629,
              "sample": "..\\..\\samples\\440.wav",
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parseSfz "https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/04%20-%20amp%20lfo%20depth%20on%20cc.sfz" 1`] = `
{
  "global": [
    {
      "group": [
        {
          "region": [
            {
              "amplfo_depthcc1": 10,
              "amplfo_freq": 5,
              "loopend": 56728,
              "loopmode": "loop_continuous",
              "loopstart": 12629,
              "sample": "..\\..\\samples\\440.wav",
              "volume": -10,
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parseSfz "https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/05%20-%20amp%20lfo%20depth%20on%20channel%20aftertouch.sfz" 1`] = `
{
  "global": [
    {
      "group": [
        {
          "region": [
            {
              "amplfo_depthchanaft": 6,
              "amplfo_freq": 5,
              "loopend": 56728,
              "loopmode": "loop_continuous",
              "loopstart": 12629,
              "sample": "..\\..\\samples\\440.wav",
              "volume": -10,
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parseSfz "https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/06%20-%20amp%20lfo%20delay.sfz" 1`] = `
{
  "global": [
    {
      "group": [
        {
          "region": [
            {
              "amplfo_delay": 0.5,
              "amplfo_depth": 3,
              "amplfo_freq": 5,
              "loopend": 56728,
              "loopmode": "loop_continuous",
              "loopstart": 12629,
              "sample": "..\\..\\samples\\440.wav",
              "volume": -10,
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parseSfz "https://raw.githubusercontent.com/sfz/tests/master/sfz1%20basic%20tests/01%20-%20Amp%20LFO/07%20-%20amp%20lfo%20fade.sfz" 1`] = `
{
  "global": [
    {
      "group": [
        {
          "region": [
            {
              "amplfo_delay": 0.5,
              "amplfo_depth": 12,
              "amplfo_fade": 1,
              "amplfo_freq": 5,
              "loopend": 56728,
              "loopmode": "loop_continuous",
              "loopstart": 12629,
              "sample": "..\\..\\samples\\440.wav",
              "volume": -12,
            },
          ],
        },
      ],
    },
  ],
}
`;
